Table SAS and SharedKey permissions

ocdi opened the pull request:
Fixes #495 and #494 and also consolidates some utility methods used across blob, table and queues.

ocdi commented:
I believe I have got this mostly complete on my draft PR, however the test cases need both fixing and fleshing out considerably. The `overrideHeaders` method in `testutils.ts` appears to be causing issues with my testing because it is modifying the `__proto__` and binding to a specific instance, which is no longer what I want it to be.

XiaoningLiu commented on src/table/middleware/AuthenticationMiddlewareFactory.ts:
```diff
@@ -0,0 +1,61 @@
+      );
+      this.authenticate(req, res, authenticators)
+        .then(pass => {
+          // TODO: To support public access, we need to modify here to reject request later in handler
```
>// TODO: To support public access, we need to modify here to reject request later in handler [](start = 10, length = 92)

Table service doesn't support public access.

ocdi replied:
A lot of my changes were copy/paste :-)

XiaoningLiu commented on src/table/utils/constants.ts:
```diff
@@ -27,3 +33,9 @@
 export const RETURN_NO_CONTENT = "return-no-content";
 export const RETURN_CONTENT = "return-content";
+
+export const EMULATOR_ACCOUNT_NAME = "devstoreaccount1";
```
>EMULATOR_ACCOUNT_NAME [](start = 13, length = 21)

We should already have emulator name and key constants in src/common/... Don't need to duplicate here.

XiaoningLiu commented on src/table/utils/utils.ts:
```diff
@@ -10,3 +13,73 @@ export function newEtag(): string {
+ * @param {string} url
+ * @returns {{[key: string]: string}}
+ */
+export function getURLQueries(url: string): { [key: string]: string } {
```
>getURLQueries [](start = 16, length = 13)

[minor] Can you help move this into src/common/utils/utils.ts? And update blob/queue/table refer to same helper method?

ocdi replied:
I did notice a fair bit of code duplication, but very happy to refactor and consolidate some of these identical methods per your suggestion here

XiaoningLiu commented on src/table/utils/utils.ts:
```diff
@@ -10,3 +13,73 @@ export function newEtag(): string {
+ *                                          If false, YYYY-MM-DDThh:mm:ssZ will be returned.
+ * @returns {string} Date string in ISO8061 format, with or without 7 milliseconds component
+ */
+export function truncatedISO8061Date(
```
>truncatedISO8061Date [](start = 16, length = 20)

[minor] Can you help move this into src/common/utils/utils.ts? And update blob/queue/table refer to same helper method?

XiaoningLiu commented on src/table/utils/utils.ts:
```diff
@@ -10,3 +13,73 @@ export function newEtag(): string {
+ * @param {key} key
+ * @returns {string}
+ */
+export function computeHMACSHA256(stringToSign: string, key: Buffer): string {
```
>computeHMACSHA256 [](start = 16, length = 17)

[minor] Can you help move this into src/common/utils/utils.ts? And update blob/queue/table refer to same helper method?

XiaoningLiu commented on src/table/errors/StorageErrorFactory.ts:
```diff
@@ -31,6 +31,67 @@ export default class StorageErrorFactory {
       context
     );
   }
+  public static getInvalidOperation(
```
>   [](start = 0, length = 2)

[Minor] Needs an addition space here.

XiaoningLiu commented on src/table/authentication/AccountSASAuthenticator.ts:
```diff
@@ -0,0 +1,316 @@
+      context.contextID
+    );
+
+    // Doesn't create a BlobContext here because wants to move this class into common
```
>BlobContext [](start = 24, length = 11)

Check typos related to blobxxx..

XiaoningLiu commented on src/table/authentication/AccountSASAuthenticator.ts:
```diff
@@ -0,0 +1,316 @@
+
+    // Doesn't create a BlobContext here because wants to move this class into common
+    const account: string = context.context.account;
+    const containerName: string | undefined = context.context.container;
```
>containerName [](start = 10, length = 13)

Should update to table concepts..

XiaoningLiu reviewed (commented):
Thanks @ocdi! Appreicated for your contribution!

Added some comments, Seems the PR is not finalized yet, will take another look when you finish.

ocdi commented:
I've addresses most of the PR comments. Added a bunch of SAS tests and solved the `__proto__` trashing issue. Need to confirm all the SAS table and account permissions are aligned correctly and fully, which is the most tedious task remaining.

XiaoningLiu commented:
Thanks @ocdi, looks great! There are some conflicts, can you have a check? after that we can merge the PR.

ocdi commented:
I rebased, the conflicts were my own work.
